### PR TITLE
feat(util): add normalizeIdentifier function

### PR DESCRIPTION
**Pull request by mttrbrts**
Adds a new function called `normalizeIdentifier` to the `@accordproject/concerto-util` package. This function allows arbitrary values to be munged into valid Concerto identifiers. 

This is useful when accepting user-provided values while dynamically building models.

### Changes

- Adds new module and function `normalizeIdentifier` and exhaustive tests

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: packages/concerto-util/lib/identifiers.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+'use strict';
+
+// Conforms to Concerto Spec for identifiers
+const ID_REGEX = /^(\p{Lu}|\p{Ll}|\p{Lt}|\p{Lm}|\p{Lo}|\p{Nl}|\$|_|\\u[0-9A-Fa-f]{4})(?:\p{Lu}|\p{Ll}|\p{Lt}|\p{Lm}|\p{Lo}|\p{Nl}|\$|_|\\u[0-9A-Fa-f]{4}|\p{Mn}|\p{Mc}|\p{Nd}|\p{Pc}|\u200C|\u200D)*$/u;
+
+/**
+ * Function that attempts to normalize arbitrary strings, numbers or booleans
+ * into valid Concerto identifiers
+ *
+ * @param {string|number|boolean} identifier - the input value
```
Comment (dselman) [inline packages/concerto-util/lib/identifiers.js:24]: Is it necessary to handle numbers and booleans? I'd lean towards forcing the client to convert them to strings to simplify this.

File: packages/concerto-util/lib/identifiers.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        .replace(/\s/g, '_')
+
+    // 3a. Replace Invalid Characters
+        .replace(/(?!\p{Lu}|\p{Ll}|\p{Lt}|\p{Lm}|\p{Lo}|\p{Nl}|\$|_|\p{Mn}|\p{Mc}|\p{Nd}|\p{Pc}|\u200C|\u200D|\\u[0-9A-Fa-f]{4})(.)/gu, replacer)
+
+    // 3b. Escape Surrogate Pairs
+        .replace(/([\uD800-\uDFFF])/g, replacer);
+
+    // 4. Optionally truncate the identifier
+    if (truncateLength > 0){
```
Comment (dselman) [inline packages/concerto-util/lib/identifiers.js:62]: truncating to 1 could make it invalid. E.g. `_`

File: packages/concerto-util/lib/identifiers.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Check validity
+    if (!ID_REGEX.test(result)){
+        throw new Error(`Unexpected error. Not able to escape identifier '${result}'.`);
+    }
+    return result;
+}
+
+module.exports = {
+    normalizeIdentifier,
```
Comment (dselman) [inline packages/concerto-util/lib/identifiers.js:74]: export `ID_REGEX` ?

**Review by dselman — COMMENTED** [review]
Couple of minor comments.

File: packages/concerto-util/lib/identifiers.js (hunk 2 above)
Comment (mttrbrts) [reply packages/concerto-util/lib/identifiers.js:62]: `_` and `$` are valid identifiers according to our spec

**Review by dselman — APPROVED** [review]
Great. Love the tests!

**Pull request merged**